libraries/AP_KDECAN/AP_KDECAN.cpp

```diff
@@ -0,0 +1,733 @@
+                    (1 << (id.source_id - ESC_NODE_ID_FIRST) & _esc_present_bitmask)) {
+                    switch (id.object_address) {
+                        case TELEMETRY_OBJ_ADDR: {
+                            if (frame.dlc != 8) {
+                                break;
+                            }
+
+                            if (!_telem_sem->take(1)) {
+                                debug_can(2, "KDECAN: failed to get telemetry semaphore on write\n\r");
+                                break;
+                            }
+
+                            _telemetry[id.source_id - ESC_NODE_ID_FIRST].time = time.toUSec();
+                            _telemetry[id.source_id - ESC_NODE_ID_FIRST].voltage = frame.data[0] << 8 | frame.data[1];
+                            _telemetry[id.source_id - ESC_NODE_ID_FIRST].current = frame.data[2] << 8 | frame.data[3];
+                            _telemetry[id.source_id - ESC_NODE_ID_FIRST].rpm = frame.data[4] << 8 | frame.data[5];
+                            _telemetry[id.source_id - ESC_NODE_ID_FIRST].temp = frame.data[6];
+                            _telemetry[id.source_id - ESC_NODE_ID_FIRST].new_data = true;
+                            _telem_sem->give();
+                            break;
+                        }
+                        default:
+                            // discard frame
+                            break;
+                    }
+                }
+            }
+        }
+
+        if (in_mask.write & inout_mask.write) {
+            now = AP_HAL::micros64();
+
+            bool new_output = _new_output.load(std::memory_order_acquire);
+
+            if (sending_esc_num > 0) {
+                if (now - pwm_last_sent > SET_PWM_TIMEOUT_US) {
+                    debug_can(2, "KDECAN: timed-out after sending frame to ESC with ID %d\n\r", sending_esc_num - 1);
+                    sending_esc_num = 0;
+                }
+            }
+
+            if (sending_esc_num == 0 && new_output) {
+                if (!_rc_out_sem->take(1)) {
+                    debug_can(2, "KDECAN: failed to get PWM semaphore on read\n\r");
+                    continue;
+                }
+
+                memcpy(output_buffer, _scaled_output, KDECAN_MAX_NUM_ESCS * sizeof(uint16_t));
+
+                _rc_out_sem->give();
+            }
+
+            if (sending_esc_num > 0 ||
+                (new_output && (pwm_last_sent == 0 || now - pwm_last_sent > SET_PWM_TIMEOUT_US)) ||
+                (pwm_last_sent != 0 && (now - pwm_last_sent > SET_PWM_MIN_INTERVAL_US))) {
+
+                for (uint8_t esc_num = sending_esc_num; esc_num < _esc_max_node_id; esc_num++) {
+
+                    if ((_esc_present_bitmask & (1 << esc_num)) == 0) {
+                        continue;
+                    }
+
+                    be16_t kde_pwm = htobe16(output_buffer[esc_num]);
+
+                    if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED) {
```

Comment:
I wonder if it's necessary to have this safety switch check this far down in the code.  It seems to me that there are already checks at a much high level (like in the AP_Motors library) that will remove the need for this.